ops/pebble.py

```diff
@@ -752,5 +861,198 @@ def get_services(self, names: List[str] = None) -> List[ServiceInfo]:
+        result = resp['result'] or []  # in case it's null instead of []
+        paths = {item['path']: item for item in result}
+        if path not in paths:
+            raise ProtocolError('path not found in response metadata: {}'.format(resp))
+        error = paths[path].get('error')
+        if error:
+            raise PathError(error['kind'], error['message'])
+
+    @classmethod
+    def _parse_read_multipart(cls, response: http.client.HTTPResponse,
+                              destinations: Dict[str, BinaryIO]) -> Dict:
+        """Parse a multipart HTTP response from the read-files API.
+
+        Return "response" metadata field decoded from JSON, and write content
+        to file-like object in destinations dictionary (keyed by path).
+        Currently the content is entirely loaded into memory, but the goal is
+        to stream that in future (the signature of read_file won't change).
+        """
+        options = cls._parse_content_type(response.headers, 'multipart/form-data')
+        boundary = options.get('boundary', '')
+        if len(boundary) < 32:
+            raise ProtocolError('invalid boundary {!r}'.format(boundary))
+
+        # We have to manually write the Content-Type with boundary, because
+        # email.parser expects the entire multipart message with headers.
+        parser = email.parser.BytesFeedParser()
+        parser.feed(b'Content-Type: multipart/form-data; boundary=' +
+                    boundary.encode('utf-8') + b'\r\n\r\n')
+
+        # Then read the rest of the response and feed it to the parser.
+        while True:
+            chunk = response.read(8192)
+            if not chunk:
+                break
+            parser.feed(chunk)
+        message = parser.close()
+
+        resp = None
+        for part in message.walk():
+            name = part.get_param('name', header='Content-Disposition')
+            if name == 'response':
+                resp = _json_loads(part.get_payload())
+            elif name == 'files':
+                # decode=True, ironically, avoids decoding bytes to str
+                content = part.get_payload(decode=True)
+                filename = part.get_filename()
+                if filename not in destinations:
+                    raise ProtocolError('path not expected: {}'.format(filename))
+                destinations[filename].write(content)
+
+        if resp is None:
+            raise ProtocolError('no "response" field in multipart body')
+
+        return resp
+
+    def write_file(
+        self, path: str, source: BinaryIO, make_dirs: bool = False, permissions: int = None,
+        user: Union[str, int] = None, group: Union[str, int] = None,
+    ):
+        """Write data from source to given file path on remote system.
+
+        If make_dirs is True, create parent directories if they don't exist.
+        Set file's mode, user, and group to those provided. User and group may
+        be either name strings or UID/GID integers.
```

Comment:
I think the general doc guideline is google docstrings with:
```
Args:
  user: user id or username for the owner of the file
  group: group id or group name for the file's group
...
```

I wouldn't block this PR on that, but we should be consistent about docstrings so that https://ops.rtfd.io works correctly

Reply:
Updated (at least the functions in this PR) to that style -- thanks.